### PR TITLE
Refactor(`lerobot_train`) Abstract Sample Weighting from RABC-Specific Implementation

### PR DESCRIPTION
**Pull request by michel-aractingi**
## Type / Scope

- **Type**: Refactor
- **Scope**: train, utils, policies/sarm

## Summary / Motivation

The training script (`lerobot_train.py`) was filled with RABC specific code in #2639. The training script should be as policy-agnostic as possible. This PR introduces a generic `SampleWeighter` abstraction that decouples sample loss weighting logic from the training script.

Sample weighting is a general training technique where individual samples are assigned weights to influence their contribution to the loss. This enables techniques like RA-BC, importance sampling, curriculum learning, and quality-based filtering. By abstracting this, the training script remains clean while still supporting these advanced training strategies.

## What changed

- **`src/lerobot/configs/train.py`**: Removed RABC-specific config parameters (`use_rabc`, `rabc_progress_path`, `rabc_kappa`, `rabc_epsilon`, `rabc_head_mode`) and replaced with a single generic `sample_weighting: SampleWeightingConfig | None` field.

- **`src/lerobot/scripts/lerobot_train.py`**: 
  - Renamed `rabc_weights_provider` parameter to `sample_weighter`
  - Replaced RABC-specific weight computation and logging with generic `SampleWeighter` interface calls
  - Removed policy-specific initialization logic (now handled by factory function)

- **`src/lerobot/utils/sample_weighting.py`** (NEW):
  - `SampleWeighter` ABC defining the interface for sample weighting strategies
  - `SampleWeightingConfig` dataclass for generic configuration
  - `make_sample_weighter()` factory function that keeps policy-specific initialization out of the training script
  - `UniformWeighter` as a no-op baseline implementation

- **`src/lerobot/policies/sarm/rabc.py`**: 
  - `RABCWeights` now inherits from `SampleWeighter` ABC
  - Added proper type hints throughout
  - Updated docstrings to reference the sample weighting infrastructure
  - Standardized stats keys (`raw_mean_weight` → `mean_weight`)

- **`tests/utils/test_sample_weighting.py`** (NEW): Comprehensive test coverage for the sample weighting infrastructure including config tests, UniformWeighter tests, factory function tests, and integration tests with RABCWeights.

## How was this tested

- Tests added: `tests/utils/test_sample_weighting.py` with 18 test cases covering:
  - `SampleWeightingConfig` initialization and validation
  - `UniformWeighter` functionality (batch size determination, device placement)
  - `make_sample_weighter` factory behavior (error handling, type dispatch)
  - Integration with `RABCWeights`

- Manual checks: Verified the refactored code maintains the same training behavior

## How to run locally (reviewer)

- Run the new tests:

  ```bash
  pytest tests/utils/test_sample_weighting.py -v
  ```

- Example config usage (for RA-BC training):

  ```bash
  lerobot-train \
    --policy.type=act \
    --dataset.repo_id=my-dataset \
    --sample_weighting.type=rabc \
    --sample_weighting.progress_path=hf://datasets/my-dataset/sarm_progress.parquet
  ```

## Checklist (required before merge)

- [x] Linting/formatting run (`pre-commit run -a`)
- [x] All tests pass locally (`pytest`)
- [x] Documentation updated
- [ ] CI is green

## Reviewer notes

- The key design decision was using an ABC (Abstract Base Class) over a Protocol. ABC was chosen for explicit inheritance and clearer error messages when methods aren't implemented.
- `SampleWeightingConfig` contains some RABC-specific fields (`progress_path`, `head_mode`, `kappa`) for convenience, but the `extra_params` dict allows future weighting strategies to pass custom parameters.
- The `UniformWeighter` is intentionally simple and serves as a baseline/no-op option.
- Breaking change: Old config parameters (`use_rabc`, `rabc_*`) are removed. Users must migrate to the new `sample_weighting` config structure.

### Migration Guide

**Before (old config):**
```python
TrainPipelineConfig(
    use_rabc=True,
    rabc_progress_path="path/to/progress.parquet",
    rabc_kappa=0.01,
    rabc_epsilon=1e-6,
    rabc_head_mode="sparse",
)
```

**After (new config):**
```python
from lerobot.utils.sample_weighting import SampleWeightingConfig

TrainPipelineConfig(
    sample_weighting=SampleWeightingConfig(
        type="rabc",
        progress_path="path/to/progress.parquet",
        kappa=0.01,
        epsilon=1e-6,
        head_mode="sparse",
    ),
)
```

### CODE REVIEW

File: src/lerobot/scripts/lerobot_train.py [hunk 1]
```diff
@@ -90,27 +90,30 @@ def update_policy(
...
     # Let accelerator handle mixed precision
     with accelerator.autocast():
-        # Use per-sample loss when RA-BC is enabled for proper weighting
-        if rabc_batch_weights is not None:
-            # Get per-sample losses
-            per_sample_loss, output_dict = policy.forward(batch, reduction="none")
+        if sample_weights is not None:
+            # Use per-sample loss for weighted training
+            # Note: Policies supporting sample weighting must implement forward(batch, reduction="none")
+            per_sample_loss, output_dict = policy.forward(batch, reduction="none")  # type: ignore[call-arg]
```
[the remark this answers is not in the document]
Comment (s1lent4gnt) [reply src/lerobot/scripts/lerobot_train.py]: Yes that's correct only Pi0, Pi05 and SmolVLA that uses SARM right now.

**Comment by s1lent4gnt** [comment]
Hey @michel-aractingi,

Thanks for the SampleWeighter work!
I cherry-picked all your commits into the reward model refactoring PR #3142. 
Made minor path adjustments to fit the new `rewards/` module layout.

Closing this since it's all incorporated.
